resolver/resolver_test.go

```diff
@@ -105,43 +131,124 @@ func TestRecurivePathResolution(t *testing.T) {
+			t.Fatal(err)
+		}
+	}
+
+	aKey := a.Cid()
+
+	fetcherFactory := bsfetcher.NewFetcherConfig(bsrv)
+	fetcherFactory.PrototypeChooser = dagpb.AddSupportToChooser(func(lnk ipld.Link, lnkCtx ipld.LinkContext) (ipld.NodePrototype, error) {
+		if tlnkNd, ok := lnkCtx.LinkNode.(schema.TypedLinkNode); ok {
+			return tlnkNd.LinkTargetNodePrototype(), nil
+		}
+		return basicnode.Prototype.Any, nil
+	})
+	fetcherFactory.NodeReifier = unixfsnode.Reify
+	r := resolver.NewBasicResolver(fetcherFactory)
+
+	// test missing link intermediate segment
+	segments := []string{aKey.String(), "cheese", "time"}
+	p, err := path.FromSegments("/ipfs/", segments...)
+	require.NoError(t, err)
+
+	_, _, err = r.ResolveToLastNode(ctx, p)
+	require.EqualError(t, err, resolver.ErrNoLink{Name: "cheese", Node: aKey}.Error())
+
+	// test missing link at end
+	bKey := b.Cid()
+	segments = []string{aKey.String(), "child", "apples"}
+	p, err = path.FromSegments("/ipfs/", segments...)
+	require.NoError(t, err)
+
+	_, _, err = r.ResolveToLastNode(ctx, p)
+	require.EqualError(t, err, resolver.ErrNoLink{Name: "apples", Node: bKey}.Error())
+}
+
+func TestResolveToLastNode_NoUnnecessaryFetching(t *testing.T) {
+	ctx := context.Background()
+	bsrv := dagmock.Bserv()
+
+	a := randNode()
+	b := randNode()
+
+	err := a.AddNodeLink("child", b)
+	require.NoError(t, err)
+
+	err = bsrv.AddBlock(a)
+	require.NoError(t, err)
+
 	aKey := a.Cid()
 
 	segments := []string{aKey.String(), "child"}
 	p, err := path.FromSegments("/ipfs/", segments...)
-	if err != nil {
-		t.Fatal(err)
-	}
+	require.NoError(t, err)
+
+	fetcherFactory := bsfetcher.NewFetcherConfig(bsrv)
+	fetcherFactory.PrototypeChooser = dagpb.AddSupportToChooser(func(lnk ipld.Link, lnkCtx ipld.LinkContext) (ipld.NodePrototype, error) {
+		if tlnkNd, ok := lnkCtx.LinkNode.(schema.TypedLinkNode); ok {
+			return tlnkNd.LinkTargetNodePrototype(), nil
+		}
+		return basicnode.Prototype.Any, nil
+	})
+	fetcherFactory.NodeReifier = unixfsnode.Reify
+	resolver := resolver.NewBasicResolver(fetcherFactory)
 
-	resolver := resolver.NewBasicResolver(dagService)
 	resolvedCID, remainingPath, err := resolver.ResolveToLastNode(ctx, p)
-	if err != nil {
-		t.Fatal(err)
-	}
+	require.NoError(t, err)
 
-	if len(remainingPath) > 0 {
-		t.Fatal("cannot have remaining path")
-	}
+	require.Equal(t, len(remainingPath), 0, "cannot have remaining path")
+	require.Equal(t, b.Cid(), resolvedCID)
+}
 
-	if !resolvedCID.Equals(b.Cid()) {
-		t.Fatal("resolved to the wrong CID")
-	}
+func TestPathRemainder(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	bsrv := dagmock.Bserv()
+
+	nb := basicnode.Prototype.Any.NewBuilder()
+	err := dagjson.Decode(nb, strings.NewReader(`{"foo": {"bar": "baz"}}`))
+	require.NoError(t, err)
+	out := new(bytes.Buffer)
+	err = dagcbor.Encode(nb.Build(), out)
+	require.NoError(t, err)
+	lnk, err := cid.Prefix{
+		Version:  1,
+		Codec:    0x71,
+		MhType:   0x17,
+		MhLength: 20,
```

Comment:
Is this intentionally off? We're using sha3-224 with 20 bytes as the length.

Reply:
This is going to make me sound like a real dolt but I don't even know what these constants mean other than the DAG-CBOR one and I don't know how long these hashes are supposed to be.

Anyway, changing to 28.

Either way, it doesn't affect the test behavior.

Reply:
We might be able to use https://github.com/multiformats/go-multicodec for more of the constants now, which may improve readability a bit.